### PR TITLE
Implement diagnostic SA1111: ClosingParenthesisMustBeOnLineOfLastParameter

### LINKED ISSUE #45: SA1111: ClosingParenthesisMustBeOnLineOfLastParameter
http://www.stylecop.com/docs/SA1111.html

### PR DESCRIPTION
**Pull request by macpak**
Fixes #45.

### CODE REVIEW

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs [hunk 1]
```diff
@@ -54,7 +58,154 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+            context.RegisterSyntaxNodeAction(HandleConstructorDeclaration, SyntaxKind.ConstructorDeclaration);
+            context.RegisterSyntaxNodeAction(HandleInvocationExpression, SyntaxKind.InvocationExpression);
+            context.RegisterSyntaxNodeAction(HandleObjectCreationExpression, SyntaxKind.ObjectCreationExpression);
+            context.RegisterSyntaxNodeAction(HandleIndexerDeclaration, SyntaxKind.IndexerDeclaration);
+            context.RegisterSyntaxNodeAction(HandleElementAccessExpression, SyntaxKind.ElementAccessExpression);
+        }
+
+        private void HandleElementAccessExpression(SyntaxNodeAnalysisContext context)
+        {
+            var tyleSymbol = context.SemanticModel.GetTypeInfo(((ElementAccessExpressionSyntax)context.Node).Expression);
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs]: A spelling mistake here

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs [hunk 2]
```diff
@@ -54,7 +58,154 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
+            if ((tyleSymbol.Type != null && tyleSymbol.Type.TypeKind != TypeKind.Array) )
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs]: Your if-statement has an extra pair of parenthesis here

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs [hunk 3]
```diff
@@ -54,7 +58,154 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
...
+
+        private static void CheckIfLocationOfLastArgumentOrParameterAndCloseTokenAreTheSame(SyntaxNodeAnalysisContext context,
+       CSharpSyntaxNode parameterOrArgument, SyntaxToken closeToken)
+        {
+            var lastParameterLine = parameterOrArgument.GetLocation().GetLineSpan();
+            var closeParenLocation = closeToken.GetLocation();
+            var closeParenLine = closeParenLocation.GetLineSpan();
+            if (lastParameterLine.IsValid &&
+                closeParenLine.IsValid &&
+                closeParenLine.StartLinePosition.Line != lastParameterLine.StartLinePosition.Line)
```
Comment (pdelvo) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs]: What happens if I write
```csharp
this.Foo("bar".
ToString())
```
Then the StartLinePositions would not match

**Comment by pdelvo** [comment]
What about Delegate declarations, anonymous delegates, lambdas? Should these also be handled by this diagnostic?

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs (hunk 3 above)
Comment (macpak) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs]: Hmm, I'd report this. What's your opinion ?

**Comment by macpak** [comment]
StyleCop description doesn't include them. The question is if WE want to include them here.

File: StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs (hunk 3 above)
Comment (sharwell) [inline StyleCop.Analyzers/StyleCop.Analyzers/ReadabilityRules/SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs]: I wouldn't report this. I read this as the closing parenthesis must be on the same line as the *last character* of the last argument.

**Comment by sharwell** [comment]
Well, since we are already considering invocations of each of these items, it would make sense to include their declarations as well. Make sure attributes get included as well.

**Comment by macpak** [comment]
Ok. We have a family of SAs which are related to one another (like #44, #47, #48). They all mention method/indexer call/declaration, the same way this one does. So I guess the same rule should apply to those issues as well - we check additonal constructs, like delegate declarations, anonymous delegates, lambdas etc.

**Comment by pdelvo** [comment]
I think so. These rules should apply to everything that Roslyn calls ParameterList or ArgumentList to be consistent.

**Comment by macpak** [comment]
Ok. So we actually make it a little bit different than in original StyleCop, I think that rule should apply to other issues - if we think that sth should be reported but is not in original StyleCop, we should add this.

**Comment by sharwell** [comment]
@macpak For now you can leave a comment in #359 if you observe a difference. Eventually we'll have a document in the repository where notes can be added.

**Comment by macpak** [comment]
Allright. Thanks for info.

**Comment by macpak** [comment]
I added code to analyze missing parts. Let me know if you think that something else should be also analyzed by this analyzer.

**Comment by sharwell** [comment]
Sweet! :sparkles: 

This one is next up on my plate for review. Plus it should lay the groundwork for many of the other readability rules. :+1:

**Comment by macpak** [comment]
Yes :) :+1:

**Comment by macpak** [comment]
I found that tests for attribute use an attribute that is not supposed to be applied on a class. I applied it on a method to make code "more correct" - actually it doesn't change anything but it's more correct :)

**Comment by sharwell** [comment]
I went ahead and merged this. So you're aware in the future, I did make a few small changes to the implementation.

1. To keep the diagnostic messages clear and short, you generally want to either duplicate the `Title` value or write a short custom message like you see in SA1310. The description is longer than users will want to read.
2. I placed the `using` directives inside the `namespace` declaration to meet the requirements of SA1200 (not yet enforced, but hopefully will be enabled by default starting with the Alpha 2 release).
3. I changed the layout of a few conditional expressions in **SA1111ClosingParenthesisMustBeOnLineOfLastParameter.cs**, which you can see by using the [blame tool on GitHub]().
4. I enabled the diagnostic by default, because once again you took the time to implement a nice set of tests. :+1:

**Comment by macpak** [comment]
Ok. Thanks.
Just one case - found that conditional expressions had bug (grrrrr) because they should've had ORs instead of combination of OR's and ANDs. For instance, these lines:

     if (lambdaExpressionSyntax.ParameterList == null ||
                !lambdaExpressionSyntax.ParameterList.IsMissing &&
                !lambdaExpressionSyntax.ParameterList.Parameters.Any())
            {
                return;
           }
should be 

    if (lambdaExpressionSyntax.ParameterList == null ||
                !lambdaExpressionSyntax.ParameterList.IsMissing ||
                !lambdaExpressionSyntax.ParameterList.Parameters.Any())
            {
                return;
           }
. I reported it on #423

**Pull request merged**